README.md
# grunt-pact-js v0.3.0-SNAPSHOT

> Generating Pact Consumer Tests with Grunt and Karma


_Note that this is not an official Grunt plugin release! If you want to use this in a project, please be sure to follow the instructions for installing development versions, as outlined in the [Installing Grunt](http://gruntjs.com/installing-grunt) guide._


## Getting Started
This plugin requires Grunt ``

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-pact-js --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-pact-js');
```

### The "contracttest" tasks

_Run this task with the `grunt contracttest` command._

This plugin relies on the Ruby pact-mock_service. Please read the chapter [Install the Mock Server](#install-the-mock-server)


### Install pact consumer library

For your tests you need [pact-consumer-js-dsl](https://github.com/DiUS/pact-consumer-js-dsl). Install the library with bower or nodejs and make it accessible over your `karmer config` file.

```
bower install pact-consumer-js-dsl --save-dev
```


```javascript
module.exports = function (config) {
    config.set({
        ...

        // list of files / patterns to load in the browser
        files: [
          // if you are using this example to setup your own project load pact from the node_modules directory
          // i.e. node_modules/pact-consumer-js-dsl/dist/pact-consumer-js-dsl.js
          'app/bower_components/pact-consumer-js-dsl/dist/pact-consumer-js-dsl.js',

        ],
        ...
    }
}
```

### Install Ruby, Gem, Bundler and Pact-MockServer

#### OS X or Linux
This task requires you to have [Ruby](http://www.ruby-lang.org/en/downloads/), [Bundler](http://bundler.io/) and [RubyGems](https://rubygems.org/pages/download) installed. If you're on OS X or Linux you probably already have Ruby installed; test with `ruby -v` in your terminal. When you've confirmed you have Ruby installed, run `gem install bundler`.

#### Windows
If on ``Windows``, please refer to the [Installing pact-mock-service on Windows](https://github.com/bethesque/pact-mock_service/wiki/Installing-the-pact-mock_service-gem-on-Windows)

Windows users must run the install command after following Wiki instructions

(This Plugin is not tested with Windows)

#### Install Mock Server

Depending on your use case you can use different commands.

##### Using the grunt plugin in a project

  Run the grunt task `grunt installMockService`.

##### Developing the plugin

 Execute `bundle install` inside the grunt pact js project folder to install pact-mock_service.

### Options

If you will change some option you can do this for example with:

```
grunt.initConfig({
        contractTest: {
            options: {
                mockServicePort: 8181,
            }
        }
    });
```

If you want change all configuration you can do this with:


```
    grunt.initConfig({
        contractTest: {
            options: {
                mockServicePort: 8181,
                karmaConfigFile: 'test/karma.conf.js',
                pactDir: 'tmp'
            }
        },
    });
```

If you want add a custom Gem file for the mock service you can do this with:

```
    grunt.initConfig({
        installMockService: {
            options: {
                gemfile: 'your-custom-file-if-you-need'
            }
        },
    });
```

The current Gem file contains the follow content:

```
    source 'https://rubygems.org'
    gem 'pact-mock_service', '~> 0.4.1'
```

It is located in `node_modules/grunt-pact-js/Gemfile`.



### Local set up

To work on this plugin locally, use the [npm link](https://docs.npmjs.com/cli/link) feature.

### Recreate README.MD

To recreate the README.MD just run `grunt build-contrib`.

### Troubleshooting

- Karma or any other task doesn't work!

```
  >> rm -rf node_modules/grunt-pact-js/node_modules/*
  >> npm install
```

- Karma still doesn't work and there were errors while npm install

  Make sure python27 is installed and selected:

```
  >> sudo port select --set python python27
  >> npm install
```

### References

See also:

- https://github.com/DiUS/pact-consumer-js-dsl
- https://github.com/bethesque/pact-mock_service/wiki/Integrating-pact-mock-service-with-Grunt-and-Protractor
- https://github.com/bethesque/pact-mock_service/wiki/Installing-the-pact-mock_service-gem-on-Windows



## Release History

 * 2015-03-27   v0.3.0-SNAPSHOT   fixed repo urls in package.json
 * 2015-03-20   v0.2.3   Remove grunt as devDependencies
 * 2015-03-20   v0.2.2   Fixed private flag and grunt-force-task dep
 * 2015-03-20   v0.2.1   Fixed typos in documentation
 * 2015-03-20   v0.2.0   Added config options
 * 2015-03-20   v0.1.0   initial setup

---

Task submitted by [ devbliss GmbH](https://www.devbliss.com/)

*This file was generated on Fri Mar 27 2015 17:48:57.*


CHANGELOG
v0.3.0-SNAPSHOT:
 date: 2015-03-27
 changes:
   - fixed repo urls in package.json
v0.2.3:
 date: 2015-03-20
 changes:
   - Remove grunt as devDependencies
v0.2.2:
 date: 2015-03-20
 changes:
   - Fixed private flag and grunt-force-task dep
v0.2.1:
 date: 2015-03-20
 changes:
   - Fixed typos in documentation
v0.2.0:
 date: 2015-03-20
 changes:
   - Added config options
v0.1.0:
 date: 2015-03-20
 changes:
   - initial setup


tasks/contract_test.js
/*
 * grunt-pact-js
 * https://github.com/devbliss/grunt-pact-js
 *
 * Copyright (c) 2015 Devbliss
 * Licensed under the Apache, License, 2.0 licenses.
 */

'use strict';

var extendGruntPlugin = require('extend-grunt-plugin');

module.exports = function (grunt) {

    grunt.registerTask('contract_test', 'Run Pact Consumer Tests with Grunt and Protractor.', function () {

            var connect = {
                options: {
                    protocol: 'http',
                    port: 8181,
                    base: '<%= build_dir %>'
                }
            };

            var shellPact = {
                command: 'node_modules/grunt-pact-js/pact/pact-mock-service-osx/pact-mock-service execute -p 1234 -l tmp/pact.log --pact-dir tmp/pacts',
                options: {
                    stdout: true,
                    stderr: true,
                    failOnError: true,
                    async: true
                }
            };

            var wait = {
                options: {
                    delay: 1000
                }
            }

            var karmaPact = {
                configFile: 'test/karma.conf.js'
            };

            extendGruntPlugin(grunt, require('grunt-shell-spawn/tasks/shell'), {
                'shell.pact': shellPact
            });

            grunt.task.run('shell:pact');

            extendGruntPlugin(grunt, require('grunt-wait/tasks/wait'), {
                'wait.pact': wait
            });

            grunt.task.run('wait:pact');

            extendGruntPlugin(grunt, require('grunt-karma/tasks/grunt-karma'), {
                'karma.pact': karmaPact
            });

            grunt.task.run('karma:pact');

            grunt.task.run('shell:pact:kill');

        }
    );
};





tasks/installMockService.js
/*
 * grunt-pact-js
 * https://github.com/devbliss/grunt-pact-js
 *
 * Copyright (c) 2015 Devbliss
 * Licensed under the Apache, License, 2.0 licenses.
 */

'use strict';

var extendGruntPlugin = require('extend-grunt-plugin');

module.exports = function (grunt) {

    grunt.registerTask('installMockService', 'This task install the Ruby Mock-Service wit Bundler and Gem.', function () {


            var options = this.options({
                gemfile: 'node_modules/grunt-pact-js/Gemfile',
            });

            var shellBundler = {
                command: 'bundle install --gemfile=' + options.gemfile,
                options: {
                    stdout: true,
                    stderr: true,
                    failOnError: true,
                    async: false
                }
            };

            extendGruntPlugin(grunt, require('grunt-shell-spawn/tasks/shell'), {
                'shell.bundler': shellBundler
            });

            grunt.task.run('shell:bundler');

        }
    );
};


tasks/contractTest.js
/*
 * grunt-pact-js
 * https://github.com/devbliss/grunt-pact-js
 *
 * Copyright (c) 2015 Devbliss
 * Licensed under the Apache, License, 2.0 licenses.
 */

'use strict';

var extendGruntPlugin = require('extend-grunt-plugin');

var contractTest = function (grunt) {

    var registerTasks = function (options) {
        var shellServerStart = {
            command: 'BUNDLE_GEMFILE="node_modules/grunt-pact-js/Gemfile"  bundle exec pact-mock-service restart -p ' +
            options.mockServicePort + ' -l  ' +
            options.pactDir + '/pact.log --pact-dir ' +
            options.pactDir + '/pacts',
            options: {
                stdout: true,
                stderr: true,
                failOnError: true,
                async: true
            }
        };

        var shellServerStop = {
            command: 'BUNDLE_GEMFILE="node_modules/grunt-pact-js/Gemfile" bundle exec pact-mock-service stop -p ' + options.mockServicePort,
            options: {
                stdout: true,
                stderr: true,
                failOnError: true,
                async: false
            }
        };

        var wait = {
            options: {
                delay: 1000
            }
        }

        var karmaPact = {
            configFile: options.karmaConfigFile,
            singleRun: true

        };


        extendGruntPlugin(grunt, require('grunt-shell-spawn/tasks/shell'), {
            'shell.serverStart': shellServerStart,
            'shell.serverStop': shellServerStop
        });

        extendGruntPlugin(grunt, require('grunt-wait/tasks/wait'), {
            'wait.pact': wait
        });

        extendGruntPlugin(grunt, require('grunt-karma/tasks/grunt-karma'), {
            'karma.pact': karmaPact
        });

        extendGruntPlugin(grunt, require('grunt-force-task/tasks/force'), {
            'force.karma': 'pact'
        });
    }

    grunt.registerTask("warning-exit", "Call process.exit", function () {
        process.exit(3);
    });

    grunt.registerTask('contractTest', 'Run Pact Consumer Tests with Grunt and Protractor.', function () {


            var options = this.options({
                mockServicePort: 1234,
                karmaConfigFile: 'test/karma.conf.js',
                pactDir: 'tmp'
            });

            registerTasks(options);

            grunt.warn = grunt.fail.warn = function (warning, callback) {
                grunt.log.warn(warning);
                grunt.task.run('shell:serverStop', 'warning-exit');
            };

            grunt.task.run('shell:serverStart', 'wait:pact', 'karma:pact', 'shell:serverStop');

        }
    );
};

module.exports = contractTest;
